# MODELİN EĞİTİM PERFORMANSINI ÖLÇME

## 1. Hava durumu modeli

Bu ödev için, geçmiş sıcaklık veri kümesine yeniden bakacaksınız.

In [2]:
import pandas as pd
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

**1. Önceki derste olduğu gibi, hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu bir doğrusal regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. R-kare ve ayarlanmış R-kare değerleri nelerdir? Tatminkar olduklarını düşünüyor musunuz? Açıklayın.**

In [2]:
Y2= weather_df["Temperature (C)"]-weather_df["Apparent Temperature (C)"]


In [3]:
X2 = weather_df[['Humidity','Wind Speed (km/h)']]

In [4]:
import statsmodels.api as sm

X3 = sm.add_constant(X2)
results = sm.OLS(Y2, X3).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sat, 01 May 2021   Prob (F-statistic):               0.00
Time:                        15:45:56   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R^2 ve adj. R^2 değerleri çok düşüktür. Model olarak kabul etmemiz içn yeterli değildir.

Model hedefteki varyansı açıklamada yetersizdir.

Model olarak kabul  edebilmemiz için R^2 en az %70 olmalıdır.

**2. Daha sonra, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Şimdi, bu modelin R-kare değeri nedir? Bu model bir öncekine göre gelişti mi?**

In [8]:
weather_df['Hum_Wind'] = weather_df.Humidity * weather_df["Wind Speed (km/h)"]

In [9]:
X6 = weather_df[['Humidity','Wind Speed (km/h)', 'Hum_Wind']]

In [10]:
import statsmodels.api as sm

X7 = sm.add_constant(X6)
results = sm.OLS(Y2, X7).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sat, 01 May 2021   Prob (F-statistic):               0.00
Time:                        16:15:47   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
Hum_Wind              0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R^2 ' de artış meydana geldi. Bu iyiye işaret ama hala yetersiz. %70 ten çok düşük

**3. İlk modele ek açıklayıcı değişken olarak görünürlük ekleyin ve tahmin edin. R-kare arttı mı? Ayarlanmış R-kare değeri ne oldu? Tabloda ortaya çıkan farklılıkları, ayarlanmış R-kare içindeki iyileşme açısından etkileşim terimi ve görünürlük ile karşılaştırın. Hangisi daha kullanışlı?**

In [13]:
X10 = weather_df[['Humidity','Wind Speed (km/h)', 'Visibility (km)']]

In [14]:
import statsmodels.api as sm

X11 = sm.add_constant(X10)
results = sm.OLS(Y2, X11).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Sat, 01 May 2021   Prob (F-statistic):               0.00
Time:                        17:00:10   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.5756      0.028    -56.605      0.000      -1.630      -1.521
Humidity              2.6066      0.025    102.784      0.000       2.557       2.656
Wind Speed (km/h)     0.1199      0.001    179.014      0.000       0.119       0.121
Visibility (km)      -0.0540      0.001    -46.614      0.000      -0.056      -0.052
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                           0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

İlk modele göre R^2 ve adj R^2 değerinde artış meydana geldi. 

Katsayıların p değerleri yaklaşık olarak sıfırdır.

Denklem içerisine Hum_Wind etkileşimini kattığımızda modelin R^2 ve adj R^2 değeri daha fazla arttı. Daha iyi bir iyileşme sağladı.

Yine de model olarak kullnamka için yetersizdir.

**4. AIC ve BIC puanlarına göre yukarıdaki üç modelden en iyisini seçin. Mentor ile gerekçenizi tartışarak seçiminizi doğrulayın.**

**1. model**

AIC:3.409e+05

BIC:3.409e+05
    
    
**2. model**

AIC:3.334e+05

BIC:3.334e+05
    
    
**3. model**

AIC:3.388e+05

BIC:3.388e+05


Her ikisi de kare hataları toplamını(SSE), örneklem boyutunu ve parametre sayısını dikkate alır.

Hem AIC hem de BIC için, daha düşük değer daha iyidir. 

3. model > 1. model > 2. model

2. model hem R^2 ve adj R^2 nin daha büyük olması hem de AIC ve BIC değerlerinin küçük olması nedeniyle daha iyidir.

## 2. Ev fiyatları modeli

Bu alıştırmada, ev fiyatları modelinizde çalışacaksınız.

In [3]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

**1. Ev fiyatları modelinizi tekrar çalıştırın ve F-testi, R-kare, ayarlanmış R-kare, AIC ve BIC kullanarak modelinizin uygunluğunu değerlendirin.**

In [6]:
Y_hp = house_prices_df['SalePrice']
X_hp= house_prices_df[['OverallQual','YearBuilt', '1stFlrSF', 'GrLivArea']]

In [5]:
import statsmodels.api as sm

X1_hp = sm.add_constant(X_hp)
results = sm.OLS(Y_hp, X1_hp).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.757
Model:                            OLS   Adj. R-squared:                  0.756
Method:                 Least Squares   F-statistic:                     1130.
Date:                Sun, 02 May 2021   Prob (F-statistic):               0.00
Time:                        12:57:32   Log-Likelihood:                -17513.
No. Observations:                1460   AIC:                         3.504e+04
Df Residuals:                    1455   BIC:                         3.506e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -9.683e+05   8.11e+04    -11.946      0.000   -1.13e+06   -8.09e+05
OverallQual  2.381e+04   1135.983     20.960      0.000    2.16e+04     2.6e+04
YearBuilt     449.9097     42.718     10.532      0.000     366.113     533.706
1stFlrSF       35.5350      3.319     10.706      0.000      29.024      42.046
GrLivArea      50.0252      2.729     18.332      0.000      44.672      55.378
==============================================================================
Omnibus:                      402.630   Durbin-Watson:                   1.995
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            33389.524
Skew:                          -0.158   Prob(JB):                         0.00
Kurtosis:                      26.426   Cond. No.                     2.19e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.19e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R^2 ve adj R^2 değeri model için iyi seviyelerde sayılır, %70 in üstündedir. 

AIC ve BIC değerleri düşük değil ama başka bir modelle karşılaştırınca daha sağlıklı  sonuç verecektir.

Ayni şekilde F statistic değeri 0 ile sonsuz arasında değer aldığı için başka model ile karşılaştırma yapmak daha sağlıklı olacaktır.



**2. Modelinizin tatmin edici olduğunu düşünüyor musunuz? Açıklayın**

R^2 ve adj R^2 değerleri  %70 in biraz üstündedir. bu yüzden iyi bir model diyebilirz.


**3.  Uygunluğunu iyileştirmek için, bazı değişkenleri ekleyerek veya kaldırarak farklı model özelliklerini deneyin.**

In [8]:
house_prices_df['Over_Gr'] = house_prices_df["OverallQual"] * house_prices_df["GrLivArea"]

In [9]:
X_hp1= house_prices_df[['OverallQual','YearBuilt', '1stFlrSF', 'GrLivArea', 'Over_Gr']]

In [10]:
import statsmodels.api as sm

X2_hp = sm.add_constant(X_hp1)
results = sm.OLS(Y_hp, X2_hp).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.773
Model:                            OLS   Adj. R-squared:                  0.772
Method:                 Least Squares   F-statistic:                     990.5
Date:                Sun, 02 May 2021   Prob (F-statistic):               0.00
Time:                        13:01:24   Log-Likelihood:                -17461.
No. Observations:                1460   AIC:                         3.493e+04
Df Residuals:                    1454   BIC:                         3.497e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -8.779e+05   7.88e+04    -11.143      0.000   -1.03e+06   -7.23e+05
OverallQual  7731.3983   1910.867      4.046      0.000    3983.048    1.15e+04
YearBuilt     458.3942     41.269     11.108      0.000     377.441     539.347
1stFlrSF       32.4213      3.220     10.068      0.000      26.105      38.738
GrLivArea     -22.2392      7.509     -2.962      0.003     -36.969      -7.510
Over_Gr        10.7523      1.046     10.278      0.000       8.700      12.804
==============================================================================
Omnibus:                     1141.250   Durbin-Watson:                   2.004
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           193270.499
Skew:                          -2.765   Prob(JB):                         0.00
Kurtosis:                      59.093   Cond. No.                     8.96e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.96e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R^2 ve adj R^2 değerinde artış gözlenmiştir. 

R^2 değeri 0.757 den 0.773 e çıkmıştır.

AIC ve BIC değerleri bir önceki modele göre düşmüştür. 

Ama F statistic azalma eğilimi göstermiştir. Model karşılaştırması yaptığımızda F statistic değerinin yüksek olması model için daha iyidir.

In [11]:
house_prices_df['Over_1st'] = house_prices_df["OverallQual"] * house_prices_df["1stFlrSF"]

In [15]:
X_hp2= house_prices_df[['OverallQual','YearBuilt', '1stFlrSF', 'GrLivArea', 'Over_1st']]

In [16]:
import statsmodels.api as sm

X3_hp = sm.add_constant(X_hp2)
results = sm.OLS(Y_hp, X3_hp).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.765
Model:                            OLS   Adj. R-squared:                  0.765
Method:                 Least Squares   F-statistic:                     949.1
Date:                Sun, 02 May 2021   Prob (F-statistic):               0.00
Time:                        13:08:12   Log-Likelihood:                -17485.
No. Observations:                1460   AIC:                         3.498e+04
Df Residuals:                    1454   BIC:                         3.501e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -9.226e+05   7.98e+04    -11.558      0.000   -1.08e+06   -7.66e+05
OverallQual  9987.3105   2168.543      4.606      0.000    5733.502    1.42e+04
YearBuilt     473.1397     42.060     11.249      0.000     390.635     555.644
1stFlrSF      -42.0974     10.941     -3.848      0.000     -63.559     -20.636
GrLivArea      50.5664      2.680     18.866      0.000      45.309      55.824
Over_1st       11.1966      1.506      7.433      0.000       8.242      14.151
==============================================================================
Omnibus:                      942.968   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           137934.274
Skew:                          -2.021   Prob(JB):                         0.00
Kurtosis:                      50.445   Cond. No.                     6.87e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.87e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**4. Denediğiniz her model için, uygun metrikleri alın ve modellerinizi birbiriyle karşılaştırın. Hangi model en iyisidir ve neden?**

'Over_Gr' eklediğimizde 'Over_1st' eklenmiş haline göre hem R^2 , adj R^2 değerleri hem de F statistic değerleri artmıştır.

Bunun yanı sıra 'Over_Gr'' modelimizde AIC ve BIC değerleri de düşmüştür. Bundan dolayı 'Over_Gr' modeli daha iyidir.